test/cbmc/stubs/network_interface_stubs.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * coreMQTT-Agent v1.0.0
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file network_interface_stubs.c
+ * @brief Stubs to mock sending and receiving over a network interface.
+ */
+
+#include "core_mqtt.h"
+#include "network_interface_stubs.h"
+
+/* An exclusive bound on the times that the NetworkInterfaceSendStub will be
+ * invoked before returning a loop terminating value. This is usually defined
+ * in the Makefile of the harnessed function. */
+#ifndef MAX_NETWORK_SEND_TRIES
+    #define MAX_NETWORK_SEND_TRIES    3
+#endif
+
+/* An exclusive bound on the times that the NetworkInterfaceReceiveStub will
+ * return an unbound value. At this value and beyond, the
+ * NetworkInterfaceReceiveStub will return zero on every call. */
+#ifndef MAX_NETWORK_RECV_TRIES
+    #define MAX_NETWORK_RECV_TRIES    4
+#endif
+
+int32_t NetworkInterfaceReceiveStub( NetworkContext_t * pNetworkContext,
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
+{
+    __CPROVER_assert( pBuffer != NULL,
+                      "NetworkInterfaceReceiveStub pBuffer is not NULL." );
+
+    __CPROVER_assert( __CPROVER_w_ok( pBuffer, bytesToRecv ),
+                      "NetworkInterfaceReceiveStub pBuffer is writable up to bytesToRecv." );
+
+    __CPROVER_havoc_object( pBuffer );
```

Comment:
I'm curious to know more about the purpose of this call. I tried to Google the function but couldn't find much documentation on it.

Reply:
I also searched for it and found only this doc http://cprover.diffblue.com/cprover__builtin__headers_8h.html. The logic for this was from the coreMQTT stub implementation. As per my understanding of the code flow and from the name of the function, the buffer will be filled with some random data which will simulate a read. So I think this is a good addition.